packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts

```diff
@@ -0,0 +1,219 @@
+			provider = getTestObjectProvider();
+			if (
+				semver.compare(provider.driver.version, "0.46.0") === -1 &&
+				(provider.driver.type === "routerlicious" || provider.driver.type === "tinylicious")
+			) {
+				this.skip();
+			}
+			const documentId = createDocumentId();
+			request = provider.driver.createCreateNewRequest(documentId);
+			loader = createTestLoader();
+		});
+
+		afterEach("resetLoaderContainerTracker", () => {
+			loaderContainerTracker.reset();
+		});
+
+		const tests = () => {
+			it("Can serialize detached container", async () => {
+				const { container } = await createDetachedContainerAndGetEntryPoint();
+
+				const snapshotTree = container.serialize();
+
+				const container2 =
+					await loader.rehydrateDetachedContainerFromSnapshot(snapshotTree);
+
+				// Check for default data store
+				const entryPoint = await container2.getEntryPoint();
+				assert.notStrictEqual(entryPoint, undefined, "Component should exist!");
+				// Might need some other asserts, but not sure what
+			});
+			it("Can serialize detached container with data stores after failed attach", async () => {
+				const { container, defaultDataStore } =
+					await createDetachedContainerAndGetEntryPoint();
+
+				// go into attaching state
+				await container.attach(request).then(
+					() => assert.fail("should fail"),
+					() => {},
+				);
+				assert.equal(
+					container.attachState,
+					AttachState.Attaching,
+					"Container should be attaching",
+				);
+
+				// create a new data store
+				const peerDataStore = await createPeerDataStore(
+					defaultDataStore.context.containerRuntime,
+				);
+				const dataStore2 = peerDataStore.peerDataStore as TestFluidObject;
+				const rootOfDataStore1 =
+					await defaultDataStore.getSharedObject<ISharedMap>(sharedMapId);
+				const dataStore2Key = "dataStore2";
+				// attach a new data store
+				rootOfDataStore1.set(dataStore2Key, dataStore2.handle);
+
+				// serialize and rehydrate
+				const snapshotTree = container.serialize();
+				const rehydratedContainer =
+					await loader.rehydrateDetachedContainerFromSnapshot(snapshotTree);
+
+				const rehydratedEntryPoint =
+					(await rehydratedContainer.getEntryPoint()) as TestFluidObject;
+				const rehydratedRootOfDataStore =
+					await rehydratedEntryPoint.getSharedObject<ISharedMap>(sharedMapId);
+
+				const dataStore2Handle: IFluidHandle<TestFluidObject> | undefined =
+					rehydratedRootOfDataStore.get(dataStore2Key);
+
+				// validate data store
+				assert(dataStore2Handle !== undefined, `handle for [${dataStore2Key}] must exist`);
+				const dataStore2FromRC = await dataStore2Handle.get();
+				assert(dataStore2FromRC, "DataStore2 should have been serialized properly");
+				assert.strictEqual(
+					dataStore2FromRC.runtime.id,
+					dataStore2.runtime.id,
+					"DataStore2 id should match",
+				);
+			});
+			it("Can serialize detached container with DDS after failed attach", async () => {
```

Comment:
can you add a test that creates both a datastore, and a dds in that datastore while attaching

Reply:
while the datastore will create dds inside, those will flow as part of the data stores attach message. the test i suggest will have an attach message for both the datastore and the dds in stashed ops